### PR TITLE
llama : fix session saving/loading

### PR DESCRIPTION
**Pull request by ggerganov**
ref #3397 

I think this should fix the issue with saving/loading session data after #3228.
Make sure to delete any old chat data

@jluisreymejias Can you give this branch a try?

### CODE REVIEW

**Comment by BarfingLemurs** [comment]
(termux) confirmed `--prompt-cache-all` + `--prompt-cache-ro` now work, while on `master` loading a created cache file led to Segmentation fault

**Comment by ggerganov** [comment]
@Senemu Could you please try your test with the latest version of this branch and see if the issue is resolved?

File: llama.h [hunk 1]
```diff
@@ -330,12 +330,16 @@ extern "C" {
             "avoid using this, it will be removed in the future, instead - count the tokens in user code");
 
     // Remove all tokens data of cells in [c0, c1)
+    // c0 < -1 : [0,  c1]
+    // c1 < -1 : [c0, inf)
```
Comment (cebtenzzre) [inline llama.h:334]: Shouldn't this be `c0 < 0`?

**Comment by ggerganov** [comment]
@Senemu I made some more changes, hoping I didn't break it again. Will merge it now without testing, but if you spot any issues again - let us know

**Comment by cebtenzzre** [comment]
> [ac2219f](https://github.com/ggerganov/llama.cpp/commit/ac2219fef34eb5b713c286c34c6e4162c39c8f3b) breaks the session cache even when using exactly the same prompt.

If this doesn't get resolved soon, open a new issue (or reopen an old one, if there is one that applies) so this doesn't get missed.

**Comment by ggerganov** [comment]
@Senemu The issue should be fixed on latest `master`.

**Pull request merged**